### ISSUE TITLE
timestamp to prevent replay attacks

### ISSUE THREAD

**Issue by boucadair** [boucadair/draft-ietf-lisp-pubsub#1]
I'm afraid that we need to have a timestamp added to the Map-Request to prevent replay attacks and thus avoid that a replayed Map-Request overwrites an existing subscription and trigger Map-Notify messages.

**Comment by arnatal** [comment]
This is why we have nonces on all control messages. We already have text for how to use the nonce to prevent replay attacks on the Map-Notify exchange, we just need to add a note regarding Map-Requests.

**Comment by boucadair** [comment]
How incremental nonces prevents replay attacks at the Map-Server?

**Comment by arnatal** [comment]
This is what RFC 9301 says for Map-Registers:

> Nonce: This 8-octet 'Nonce' field is incremented each time a Map-Register message is sent. When a Map-Register acknowledgment is requested, the nonce is returned by Map-Servers in Map-Notify messages. Since the entire Map-Register message is authenticated, the 'Nonce' field serves to protect against Map-Register replay attacks. An ETR that registers to the Mapping System SHOULD store the last nonce sent in persistent storage, so when it restarts, it can continue using an incrementing nonce. If the ETR cannot support saving the nonce, then when it restarts, it MUST use a new authentication key to register to the Mapping System. A Map-Server MUST track and save in persistent storage the last nonce received for each ETR xTR-ID and key pair. If a Map-Register is received with a nonce value that is not greater than the saved nonce, it MUST drop the Map-Register message and SHOULD log the fact that a replay attack could have occurred

The PubSub docs follows this to prevent replay attacks for Map-Notify used as publications:

> If a Map-Notify received as a publication has a nonce value that is not greater than the saved nonce, the xTR drops the Map-Notify message and logs the fact a replay attack could have occurred. The same considerations discussed in Section 5.6 of [[RFC9301](https://www.ietf.org/archive/id/draft-ietf-lisp-pubsub-10.html#RFC9301)] regarding Map-Register nonces apply here for Map-Notify nonces.

We would still need to recommend authentication / integrity protection between the xTR and the MR/MS in deployments that could be subject to replay attacks (this aligns with the guidelines on 9301/9303).

**Comment by boucadair** [comment]
Fully agree on the auth/integrity protection reco. Also, agree that the nonce++ is a good guard against replay attacks of Map-Notify at the ETR. 

However, I'm not sure we are good when it comes to Map-Requests:

RFC9301 says simply:

>    Nonce:  This is an 8-octet random value created by the sender of the
>       Map-Request.  This nonce will be returned in the Map-Reply.  The
>       nonce is used as an index to identify the corresponding Map-
>       Request when a Map-Reply message is received.  The nonce MUST be
>       generated by a properly seeded pseudo-random source; for example,
>       see [RFC4086].

Pubsub does not enforce any check on the nonce to override an existing state: 

>    If an xTR-ID is successfully added to the list of subscribers for an
>    EID-Record, the Map-Server MUST extract the nonce and ITR-RLOCs
>    present in the Map-Request, and store the association between the
>    EID-Record, xTR-ID, ITR-RLOCs and nonce.  Any already present state
>    regarding ITR-RLOCs and/or nonce for the same xTR-ID MUST be
>    overwritten.

So a replayed Map-Request will pass through.

I think that the situation is even worse when it comes to withdrawal as no check is made on the nonce to use. A replayed Map-Request will always remove installed state:

>    The following specifies the procedure to remove a subscription.  If
>    the Map-Request only has one ITR-RLOC with AFI = 0 (i.e., Unknown
>    Address), the Map-Server MUST remove the subscription state for that
>    xTR-ID.  In this case, the Map-Server MUST send the Map-Notify to the
>    source RLOC of the Map-Request.

**Comment by arnatal** [comment]
Agreed, we're not good as of now. We need to add some text to mention that the nonce should be checked on Map-Request as well. I think that along with the recommendation on Map-Resolvers checking auth/integrity should be enough.

**Comment by boucadair** [comment]
Great, thanks...but

Relying on an incremental nonce will have a side effect for amnesic xTRs  (no persistent storage of nonce) or during maintenance operations without state sync. The xTR can't subscribe since its requests will be discarded because of the hard checks on the nonce. Also, a new xTR can't clear any stale state when it first bootstrapps. We can't guarantee that xtr_ids won't be recycled, etc. 

I'm afraid that a mix of nonce++/timestamp si needed:
* timestamp for first subscription then nonce++/timestamp for updates
* no check on the nonce but require a timestamp for deletion

What do you think?

**Comment by arnatal** [comment]
Yes, the issue with amnesic xTRs was discussed during the 9301 standardisation process. As outcome, RFC9301 now reads:

> A Map-Server MUST track and save in persistent storage the last nonce

Which doesn't leave really much room for amnesic xTRs. I think we should align with 9301.

**Comment by boucadair** [comment]
The same para from where you extracted the quote also says: 

> An ETR that registers to the Mapping System SHOULD store
>       the last nonce sent in persistent storage, so when it restarts, it
>       can continue using an incrementing nonce.  If the ETR cannot
>       support saving the nonce, then when it restarts, it MUST use a new
>       authentication key to register to the Mapping System.

Isn't odd to have a MUST for xTR for nonce storage for Map-Requests, while this is only a SHOULD for Map-Register?

**Comment by arnatal** [comment]
Sorry I had in mind the MS, you’re right that we ended up with a SHOULD for ETRs and Map-Register. If we want to support a SHOULD also for ITRs and Map-Requests, one option is indeed to explore the timestamp route. Unfortunately, that would depart from the rest of the LISP architecture where timestamps are not used (afaik). I wonder if we could find a way to handle this that does not require to introduce a new concept in LISP (also we prevent unexpected side effects, if any). 

Thinking on this, I wonder if the solution might be on Section 7.1 of PubSub. There we describe a tamper-proof way to subscribe and create a security association (provided the security guidelines of PubSub, 9301, and 9303 are followed). How about we specify that for scenarios subject to replay attacks that include ITRs that lose state (non-persistent storage) the procedure described in Section 7.1 MUST be followed to restart the nonce (the MS needs to check that the OTK that restarts the nonce has been not used before of course). This would be equivalent to ETRs on 9301 that require to use a new authentication key when persistent storage is not an option. What do you think?

**Comment by boucadair** [comment]
> Sorry I had in mind the MS, you’re right that we ended up with a SHOULD for ETRs and Map-Register. If we want to support a SHOULD also for ITRs and Map-Requests, one option is indeed to explore the timestamp route. Unfortunately, that would depart from the rest of the LISP architecture where timestamps are not used (afaik). I wonder if we could find a way to handle this that does not require to introduce a new concept in LISP (also we prevent unexpected side effects, if any).
> 

ACK. This is not used currently but pubsub have specifics. However, it is pragmatic to explore first the other option below. 

> Thinking on this, I wonder if the solution might be on Section 7.1 of PubSub. There we describe a tamper-proof way to subscribe and create a security association (provided the security guidelines of PubSub, 9301, and 9303 are followed). 

I assume this will cover also adding an explicit mention that we deviate from 5.2 of 9301 for nonce generation for sub updates (for whatever reason local to the xTR): The nonce++ of Map-Request for a specific EID used by the xTR (updates) should be based on the nonce seen in the latest (authenticated) map-notify. 

> How` about we specify that for scenarios subject to replay attacks that include ITRs that lose state (non-persistent storage) the procedure described in Section 7.1 MUST be followed to restart the nonce (the MS needs to check that the OTK that restarts the nonce has been not used before of course). This would be equivalent to ETRs on 9301 that require to use a new authentication key when persistent storage is not an option. What do you think?

The MS will accept an update/withdrawal, if nonce++ or new distinct nonce but with a new authentication key. I think this would work but this has scalability issues as the MS has to store every key-id used in a Map-Request. Otherwise, the replayed message can pass through when when that key was removed from the MS logs.  I expect also comments about how to ensure that a distinct key is used when restarts especially for xTR without permanent storage, though.  

BTW, I started https://github.com/boucadair/lisp-pubsub-flow-examples to exercise the spec and illustrate the expected behavior. The current version assumes the nonce++/key-id storage approach.

**Comment by boucadair** [comment]
I think that nonce++ still don't solve the replay attack issue when the replayed attack is sent after a withdrawal. to make it short, nonce++ works fine when there is a state in the Map-Server ... but is broken when no state is maintained by the Map-Server.

**Comment by arnatal** [comment]
As we discussed above, the Map-Server MUST keep the nonce in persistent storage for Map-Registers in 9301. The same can be used here, even to keep the nonce after a withdrawal.

Please keep also in mind that many production deployments of LISP run a single logical Map-Server, where a single nonce can be kept per xTR-ID (as we describe in the PubSub spec).

**Comment by boucadair** [comment]
> even to keep the nonce after a withdrawal.

I'm not comfortable to say that in a spec...

> where a single nonce can be kept per xTR-ID (as we describe in the PubSub spec).

This one softens (but does not nullify) the issue if an xTR has at least other active state. Worth be mentioned.

**Comment by arnatal** [comment]
I pushed a PR that tries to find a good compromise and fall backs to keeping/verifying OTKs, when nonces are gone. I also added some text to mention that a hash can be used to keep the "previous OTKs" state easier to manage. A explicit auth-failed Map-Replay is sent now when the OTK check fails, so a balance can be achieved between the hash length and how common is that a xTR needs to retry due to hash collisions. I hope this achieves a good balance overall, let me know what you think.
